script/libraries/LibInitializeGuard.sol

```diff
@@ -0,0 +1,343 @@
+
+/**
+ * @dev Library to guard the initialization of the proxies and logics.
+ * - Proxy:
+ *   + The proxy MUST have `_initialized` slot.
+ *   + `_initialized` value MUST increment by 1 after each `initialize` function call.
+ *   + The last initialized version MUST equal to the number of `initialize` functions.
+ * - Logic:
+ *   + The logic MUST disable the initialized version.
+ */
+library LibInitializeGuard {
+  using StdStyle for *;
+  using LibString for string;
+  using JSONParserLib for string;
+  using JSONParserLib for JSONParserLib.Item;
+  using EnumerableSet for EnumerableSet.AddressSet;
+
+  struct InitLocation {
+    bytes32 slot;
+    uint256 bitOffset;
+    uint256 nBit;
+  }
+
+  struct Cache {
+    EnumerableSet.AddressSet _logics;
+    EnumerableSet.AddressSet _proxies;
+    mapping(address addr => uint256) _lastInitVer;
+    mapping(address addr => Vm.ChainInfo) _chainInfo;
+    mapping(address proxy => InitLocation) _initSlot;
+    mapping(address logic => address proxy) _logic2Proxy;
+  }
+
+  /// @dev Number of bits used to store initialized version in `_initialized` slot of OpenZeppelin v4
+  uint256 private constant N_BIT_INIT_V4 = 8;
+  /// @dev Maximum value of initialized version in `_initialized` slot of OpenZeppelin v4
+  uint256 private constant MAX_VER_V4 = type(uint8).max;
+  /// @dev Number of bits used to store initialized version in `_initialized` slot of OpenZeppelin v5
+  uint256 private constant N_BIT_INIT_V5 = 64;
+  /// @dev Maximum value of initialized version in `_initialized` slot of OpenZeppelin v5
+  uint256 private constant MAX_VER_V5 = type(uint64).max;
+
+  /// @dev See: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.0.2/contracts/proxy/utils/Initializable.sol#L77
+  bytes32 private constant INITIALIZABLE_STORAGE_OZV5 =
+    0xf0c57e16840df040f15088dc2f81fe391c3923bec73e23a9662efc9c229c6a00;
+
+  /// @dev Custom storage slot of the `Cache` struct
+  bytes32 private constant $$_CacheStorageLocation = keccak256("LibInitializeGuard.Cache.storage.slot");
+  /// @dev Custom storage slot of the `StdStorage` struct
+  bytes32 private constant $$_StdStorageLocation = keccak256("LibInitializeGuard.StdStorage.storage.slot");
+
+  /**
+   * @dev Validate the initialization of the proxies and logics.
+   *
+   * Requirements:
+   * - Must record `logs` via `vm.recordLogs()` before calling this function.
+   * - Must record `stateDiffs` via `vm.startStateDiffRecording()` before calling this function.
+   *
+   * @param logs The logs of the transactions.
+   * @param stateDiffs The state diffs of the transactions.
+   */
+  function validate(Vm.Log[] memory logs, Vm.AccountAccess[] memory stateDiffs) internal {
+    Cache storage $ = _getCacheStorage();
+
+    _recordUpgradesAndInitializations({ $cache: $, logs: logs });
+
+    for (uint256 i; i < stateDiffs.length; ++i) {
+      address addr = stateDiffs[i].account;
+
+      if ($._proxies.contains(addr) && $._initSlot[addr].nBit != 0) {
```

Comment:
```suggestion
      if ($._proxies.contains(addr)) {
```
I found a weird bug right here when writing test, inside `_recordUpgradesAndInitializations` does not update initSlot.nBit, so this condition will always be false even if there is a valid OpenZeppelin v5 `Initializable` contract. I think checking if addr is contained in $._proxies is enough

Reply:
fixed. second condition is to skip writing if `initSlot` is already visited for given proxy.